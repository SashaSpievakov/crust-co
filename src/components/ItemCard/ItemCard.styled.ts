import styled from 'styled-components';
import { Link } from 'react-router-dom';

const Block = styled.article`
  width: 250px;
  text-align: center;
  margin-bottom: 65px;

  &:hover {
    img {
      transform: translateY(-10px);
    }

    h4 {
      color: ${({ theme }) => theme.colors.secondary};
    }
  }

  @media screen and (max-width: 400px) {
    width: 200px;
  }
`;

const CustomLink = styled(Link)`
  display: block;
  border-radius: 5px;

  &:focus-visible {
    outline: 5px solid ${({ theme }) => theme.colors.third};
  }
`;

const Image = styled.img`
  margin-bottom: 15px;
  border-radius: 20px;
  transition: 0.2s ease-in-out;

  @media screen and (max-width: 400px) {
    width: 200px;
    height: 200px;
  }
`;

const Title = styled.h3`
  font-size: 20px;
  font-weight: 900;
  letter-spacing: 1%;
  margin-bottom: 20px;
`;

const Bottom = styled.section`
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  margin-top: 20px;
`;

const Price = styled.h3`
  font-weight: bold;
  font-size: 32px;
  line-height: 27px;
  letter-spacing: 0.015em;

  @media screen and (max-width: 400px) {
    font-size: 28px;
  }
`;

export { Block, CustomLink, Image, Title, Bottom, Price };
