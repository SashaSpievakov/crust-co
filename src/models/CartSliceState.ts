import { ICartItem } from './ICartItem';

export interface CartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
