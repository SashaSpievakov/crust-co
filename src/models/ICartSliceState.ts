import { ICartItem } from './ICartItem';

export interface ICartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
