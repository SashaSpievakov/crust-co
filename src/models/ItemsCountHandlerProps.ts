export interface ItemsCountHandlerProps {
  id: string;
  name: string;
  price: number;
  activeSize: number;
  activeType: number;
  typeNames: string[];
  isFullScreen?: boolean;
}
