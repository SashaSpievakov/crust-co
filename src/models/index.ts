export type { ICartItem } from './ICartItem';
export type { IPizzaItem } from './IPizzaItem';
