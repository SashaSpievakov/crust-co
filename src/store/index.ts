export * from './slices';
export type { AppDispatch, AppStore, RootState } from './store';
export { setupStore } from './store';
