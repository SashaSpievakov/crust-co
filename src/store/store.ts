import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import categorySlice from './slices/category/reducer/categoryReducer';
import sortSlice from './slices/sort/reducer/sortReducer';
import searchSlice from './slices/search/reducer/searchValueReducer';
import theme from './slices/theme/reducer/themeReducer';
import cart from './slices/cart/reducer/cartReducer';
import itemAPI from '../services/ItemService';
import pizzasAPI from '../services/PizzasService';

const rootReducer = combineReducers({
  activeCategory: categorySlice,
  activeSort: sortSlice,
  searchValue: searchSlice,
  theme,
  cart,
  [itemAPI.reducerPath]: itemAPI.reducer,
  [pizzasAPI.reducerPath]: pizzasAPI.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(itemAPI.middleware)
        .concat(pizzasAPI.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
