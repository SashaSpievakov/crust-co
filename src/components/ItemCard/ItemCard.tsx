import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';

import ItemsCountHandler from '../UI/ItemsCountHandler/ItemsCountHandler';
import Selector from '../UI/Selector/Selector';
import { Block, Image, Title, Bottom, Price } from './ItemCard.styled';

interface ItemCardProps {
  id: string;
  name: string;
  price: number;
  sizes: number[];
  types: number[];
}

const typeNames: string[] = ['traditional', 'thin'];

const ItemCard = ({ id, name, price, sizes, types }: ItemCardProps) => {
  const [activePrice, setActivePrice] = useState<number>(price);
  const [activeSize, setActiveSize] = useState<number>(sizes[0]);
  const [activeType, setActiveType] = useState<number>(0);

  useEffect(() => {
    setActivePrice(price);
  }, [sizes]);

  return (
    <Block>
      <Link to={`/item/${id}`}>
        <picture>
          <source srcSet={`./assets/img/pizza${id}.webp`} type="image/webp" />
          <source srcSet={`./assets/img/pizza${id}.png`} type="image/png" />
          <Image
            src={`./assets/img/pizza${id}.png`}
            alt={`${name} pizza`}
            width="250"
            height="250"
          />
        </picture>
        <Title>{name}</Title>
      </Link>
      <Selector
        price={activePrice}
        sizes={sizes}
        types={types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
      />
      <Bottom>
        <Price>{activePrice}$</Price>
        <ItemsCountHandler
          id={id}
          name={name}
          price={activePrice}
          activeSize={activeSize}
          activeType={activeType}
          typeNames={typeNames}
        />
      </Bottom>
    </Block>
  );
};

export default ItemCard;
