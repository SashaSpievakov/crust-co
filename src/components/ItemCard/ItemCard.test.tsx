import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';

import ItemCard from './ItemCard';
import rendererWithAllProviders from '../../tests/helpers/rendererWithProviders';
import renderWithAllProviders from '../../tests/helpers/renderWithProviders';
import { IPizzaItem } from '../../models/IPizzaItem';
import { setupStore } from '../../store/store';
import server from '../../tests/mocks/api/server';
import itemAPI from '../../services/ItemService';

const ItemCardProps: IPizzaItem = {
  id: '9',
  name: 'Vegetarian Pizza',
  description:
    'Lorem, ipsum dolor sit amet consectetur adipisicing elit. Aliquid nisi aspernatur debitis quod suscipit, vitae placeat natus ipsum inventore esse rerum animi facere numquam saepe vero mollitia quibusdam at voluptates est commodi laudantium? Iste inventore quibusdam cupiditate nemo sint iusto minus nihil culpa! Sequi architecto nesciunt explicabo mollitia. Laboriosam, odio.',
  types: [0, 1],
  sizes: [12, 14, 16],
  price: 8,
  category: 3,
  rating: 15,
};

describe('ItemCard Test', () => {
  test('renders the ItemCard component', () => {
    const snapshot = rendererWithAllProviders(<ItemCard {...ItemCardProps} />);
    expect(snapshot).toMatchSnapshot();
  });

  describe('checks the price changing', () => {
    beforeEach(() => {
      renderWithAllProviders(<ItemCard {...ItemCardProps} />);
    });

    test('checks type click', () => {
      const typeItem = screen.getByText(/thin/i);
      const price = screen.getByText(/8\$/i);

      userEvent.click(typeItem);

      expect(price).toHaveTextContent('9$');
    });

    test('checks size click', () => {
      const sizeItem = screen.getByText(/16 inch/i);
      const price = screen.getByText(/8\$/i);

      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('12$');
    });

    test('checks type and size click', () => {
      const typeItem = screen.getByText(/thin/i);
      const sizeItem = screen.getByText(/14 inch/i);
      const price = screen.getByText(/8\$/i);

      userEvent.click(typeItem);
      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('11$');
    });
  });

  describe('checks rtk query hook', () => {
    beforeAll(() => {
      server.listen();
    });

    afterEach(() => {
      server.resetHandlers();
      setupStore().dispatch(itemAPI.util.resetApiState());
    });

    afterAll(() => {
      server.close();
    });

    test('checks link redirect to the Item page', async () => {
      server.use(
        rest.get('*', (req, res, ctx) => {
          return res(ctx.json(ItemCardProps));
        }),
      );

      window.HTMLElement.prototype.scrollIntoView = function () {}; // eslint-disable-line func-names
      renderWithAllProviders(<ItemCard {...ItemCardProps} />, true, '/cart');
      const link = screen.getByRole('heading', {
        name: /vegetarian pizza/i,
      });

      userEvent.click(link);

      await waitFor(() => {
        expect(
          screen.getByText(
            /Lorem, ipsum dolor sit amet consectetur adipisicing elit. Aliquid nisi aspernatur debitis quod/i,
          ),
        ).toBeInTheDocument();
      });
    });
  });
});
