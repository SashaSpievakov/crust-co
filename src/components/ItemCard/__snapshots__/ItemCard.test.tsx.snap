// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard Tests renders the ItemCard component 1`] = `
Object {
  "children": Array [
    <a
      className="sc-ksBlkl fhhtjn"
      href="/item/9"
      onClick={[Function]}
    >
      <picture>
        <source
          srcSet="./assets/img/pizza9.webp"
          type="image/webp"
        />
        <source
          srcSet="./assets/img/pizza9.png"
          type="image/png"
        />
        <img
          alt="Chicken Curry pizza"
          className="sc-hBxehG hiutnt"
          height="250"
          src="./assets/img/pizza9.png"
          width="250"
        />
      </picture>
      <h3
        className="sc-fnGiBr djgFrE"
      >
        Chicken Curry
      </h3>
    </a>,
    <section
      className="sc-csuSiG fpLAMx"
    >
      <ul>
        <li
          aria-current={true}
          className="sc-eDWCr izhotF"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          traditional
        </li>
        <li
          aria-current={false}
          className="sc-eDWCr ksHZVK"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          thin
        </li>
      </ul>
      <ul>
        <li
          aria-current={true}
          className="sc-eDWCr izhotF"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          12
           inch
        </li>
        <li
          aria-current={false}
          className="sc-eDWCr ksHZVK"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          14
           inch
        </li>
        <li
          aria-current={false}
          className="sc-eDWCr ksHZVK"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          16
           inch
        </li>
      </ul>
    </section>,
    <section
      className="sc-fEXmlR dJUDNk"
    >
      <h3
        className="sc-bjfHbI dFHOIQ"
      >
        14
        $
      </h3>
      <button
        className="sc-bcXHqe sc-eDvSVe sc-gKPRtg glVyWT fwWQTF gmBSoV"
        onClick={[Function]}
      >
        <svg
          className="sc-ipEyDJ fbIxLm"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
          />
        </svg>
        <span>
          Add
        </span>
      </button>
    </section>,
  ],
  "props": Object {
    "className": "sc-bqWxrE fptGBs",
  },
  "type": "article",
}
`;
